src/checkpoint/dispatch/vrt/derived.h

```diff
@@ -0,0 +1,124 @@
+//   software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact darma@sandia.gov
+//
+// *****************************************************************************
+//@HEADER
+*/
+
+#if !defined INCLUDED_CHECKPOINT_DISPATCH_VRT_DERIVED_H
+#define INCLUDED_CHECKPOINT_DISPATCH_VRT_DERIVED_H
+
+#include "checkpoint/dispatch/vrt/registry_common.h"
+#include "checkpoint/dispatch/vrt/static_dispatch_typeidx.h"
+#include "checkpoint/dispatch/vrt/object_registry.h"
+#include "checkpoint/dispatch/vrt/serializer_registry.h"
+#include "checkpoint/dispatch/vrt/inheritance_assert_helpers.h"
+#include "checkpoint/dispatch/vrt/serialize_instantiator.h"
+
+#define checkpoint_virtual_serialize_derived(DERIVED, BASE)                          \
+  void _checkpointDynamicSerialize(                                                  \
+    void* s,                                                                         \
+    ::checkpoint::dispatch::vrt::TypeIdx base_ser_idx,                               \
+    ::checkpoint::dispatch::vrt::TypeIdx expected_idx                                \
+  ) override {                                                                       \
+    ::checkpoint::instantiateObjSerializer<                                          \
+      DERIVED,                                                                       \
+      checkpoint_serializer_variatic_args()                                          \
+    >();                                                                             \
+    debug_checkpoint(                                                                \
+      "%s: BEGIN: _checkpointDynamicSerialize: serializer_idx=%d {\n",               \
+      typeid(DERIVED).name(), base_ser_idx                                           \
+    );                                                                               \
+    ::checkpoint::dispatch::vrt::assertTypeIdxMatch<DERIVED>(expected_idx);          \
+    auto base_idx = ::checkpoint::dispatch::vrt::objregistry::makeObjIdx<BASE>();    \
+    BASE::_checkpointDynamicSerialize(s, base_ser_idx, base_idx);                    \
```

Comment:
Do I need some extra disambiguation here if BASE is actually a more complex type with template parameters? What happens when it's not?

Reply:
I think because you're ultimately calling a function, all should be well. I can see how we might get into a weird space where `typename` is hypothetically required of us for some uses, and forbidden for others.

Or is your concern that `DERIVED` and/or `BASE` might contain commas in template arguments, thus sans parentheses, and thereby confuse the preprocessor about how many arguments are being passed to the macro? We may need to require and document that arguments to these macros have to be either templates of just one argument, or a typedef to a single symbol, to avoid that problem.

Reply:
I was thinking about the typename issue, but I think the commas are potentially an issue also. Requiring a typedef in those cases seems reasonable. Extra parens will make C++ unhappy here.

A typedef can also solve the potential typename ambiguity issue also.

Reply:
On the checkpoint end, I should add comments about this *and* in `Serializable{Base,Derived}` make a typedef before calling the macro.